### PR TITLE
Refactor `NodePropertyUpdate` in preparation for composite indexes.

### PR DESCRIPTION
**Pull request by ragadeeshu**
Node changes now manifests as `NodeUpdate` objects, that contain the full set of properties and labels for the node, before and after the change. From the `NodeUpdate` we can find the changes that are relevant to a particular index using `forIndex()`, which returns an `IndexEntryUpdate`. The `IndexEntryUpdate` contains only the information of interest to one particular index.

This solution solution can likely cause performance degradation as we are looking up more properties that before, causing more db hits. We will try to alleviate this later by only looking up properties that are known to be of interest for some index.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/NodeUpdates.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    }
+
+    public final long getNodeId()
+    {
+        return nodeId;
+    }
+
+    private final Map<Integer,Object> propertiesBefore = new HashMap<>();
+    private final Map<Integer,Object> propertiesAfter = new HashMap<>();
+    private final Map<Integer,Object> propertiesUnchanged = new HashMap<>();
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/NodeUpdates.java]: Please move these fields up to the other fields.

File: community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class IndexEntryUpdateTest
+{
+    @Test
+    public void indexEntryUpdatesShouldBeEqual()
+    {
+        IndexEntryUpdate a = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), "hi" );
+        IndexEntryUpdate b = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), "hi" );
+        assertThat( a, equalTo( b ) );
+        assertThat( a.hashCode(), equalTo( b.hashCode() ) );
+    }
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java:47]: I'm sure there are more cases to test than just this one.

File: community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java [hunk 3]
```diff
@@ -935,7 +935,7 @@ public void shouldRunIndexPopulationJobAtShutdown() throws Throwable
         verify( provider ).getPopulator( anyLong(), any( IndexDescriptor.class ), any( IndexConfiguration.class ),
                 any( IndexSamplingConfig.class ) );
         verify( populator ).create();
-        verify( populator ).add( singletonList( NodePropertyUpdate.add( nodeId, 0, "Jakewins", new long[]{0} ) ) );
+        verify( populator ).add( singletonList( IndexEntryUpdate.add( nodeId, any(), "Jakewins" ) ) );
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java]: Are you sure these `any()` calls mean what you think it means?

**Review by chrisvest — COMMENTED** [review]
I had some comments.

**Comment by ragadeeshu** [comment]
@chrisvest Please take a look again.

File: community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexEntryUpdate.java [hunk 4]
```diff
@@ -85,23 +85,44 @@ public boolean equals( Object o )
...
+        // Probably incorrect - comparing Object[] arrays with Arrays.equals
+        if ( !Arrays.equals( before, that.before ) )
+        {
+            return false;
+        }
+        // Probably incorrect - comparing Object[] arrays with Arrays.equals
+        if ( !Arrays.equals( values, that.values ) )
+        {
+            return false;
+        }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexEntryUpdate.java]: Why are these probably incorrect?

File: community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java [hunk 5]
```diff
@@ -36,4 +45,76 @@ public void indexEntryUpdatesShouldBeEqual()
...
+        IndexEntryUpdate singleChange = IndexEntryUpdate
+                .change( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), singleValue, singleAfter );
+        Object[] multiAfter = {"Hello", "Hi"};
+        IndexEntryUpdate multiChange = IndexEntryUpdate
+                .change( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), multiValue, multiAfter );
+        assertThat( new Object[]{singleValue}, equalTo( singleChange.beforeValues() ) );
+        assertThat( new Object[]{singleAfter}, equalTo( singleChange.values() ) );
+        assertThat( multiValue, equalTo( multiChange.beforeValues() ) );
+        assertThat( multiAfter, equalTo( multiChange.values() ) );
+    }
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java:119]: _Magnifique!_

File: community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java [hunk 6]
```diff
@@ -939,7 +939,8 @@ public void shouldRunIndexPopulationJobAtShutdown() throws Throwable
         verify( provider ).getPopulator( anyLong(), any( IndexDescriptor.class ), any( IndexConfiguration.class ),
                 any( IndexSamplingConfig.class ) );
         verify( populator ).create();
-        verify( populator ).add( singletonList( IndexEntryUpdate.add( nodeId, any(), "Jakewins" ) ) );
+        verify( populator ).add( singletonList( IndexEntryUpdate.add( nodeId, any( NewIndexDescriptor.class)
+                , "Jakewins" ) ) );
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java]: And this we discussed in the office.

File: community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java [hunk 7]
```diff
@@ -117,4 +117,21 @@ public void changedShouldRetainValues() throws Exception
...
+
+        IndexEntryUpdate a = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), value1 );
+        IndexEntryUpdate b = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), value2 );
+        assertThat( a, equalTo( b ) );
+        assertThat( a.hashCode(), equalTo( b.hashCode() ) );
+    }
+
+    private Object[] arraysOfDepth( int n )
+    {
+        return n == 0 ? new Object[]{"hej"} : arraysOfDepth( n-1 );
```
Comment (chrisvest) [inline community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java]: This doesn't wrap or nest any arrays. It just returns the array at the recursion base case.

**Review by chrisvest — APPROVED** [review]

**Pull request merged**